Rewrite appearance for 2025 (#40564)
mdn/content: files/en-us/web/css/appearance/index.md

The `base-select` value, which is relevant only to the {{htmlelement("select")}} element and {{cssxref("::picker()", "::picker(select)")}} pseudo-element, enables [styling `<select>` elements and the select picker](#Setting_the_appearance_of_a_select) (which contains the `<option>` elements). The picker is rendered in the top layer, similar to a popover. When `base-select` is set, the picker can be positioned relative to the select (or other elements) using [CSS anchor positioning](/en-US/docs/Web/CSS/CSS_anchor_positioning) features. In addition, the `base-select` value causes the `<select>` not to render outside the browser pane or to trigger built-in mobile operating system components. It is also no longer sized based on the width of the widest `<option>`.

### Prefixed non-standard values

> Before standardization, the prefixed **`-moz-appearance`** and **`-webkit-appearance`** properties allowed elements to be shown as widgets such as buttons or checkboxes. The following non-standard values may be encountered in legacy style sheets, most commonly as values of shadow DOM component [prefixed pseudo-elements](/en-US/docs/Web/CSS/WebKit_Extensions#pseudo-elements):

Author: ```suggestion
Before standardization, the prefixed **`-moz-appearance`** and **`-webkit-appearance`** properties allowed elements to be shown as widgets such as buttons or checkboxes. The following non-standard values may be encountered in legacy style sheets, most commonly as values of shadow DOM component [prefixed pseudo-elements](/en-US/docs/Web/CSS/WebKit_Extensions#pseudo-elements).
```

While most elements in a document can be fully controlled by CSS, UI controls, or _widgets_, are typically rendered by the browser using the operating system's native UI controls. This _native_ appearance differs between operating systems and browsers. In this default state, widgets have limited, if any, features that can be styled with CSS. Which elements have this native UI appearance is defined in HTML.

The `appearance` property provides some control over the appearance of HTML widgets that by default look like native operating system controls. Most notably, the `none` value suppresses some of a widget's native appearance, giving the widget a _primitive_ appearance, enabling the widget to be styled via CSS while still maintaining the functionality and supporting native user interactions.

Widgets with a primitive appearance have their native appearance suppressed, causing some widgets types to disappear completely. While visually hidden, with inline, block, border, and padding sizes of `0`, the hidden controls are still interactive. For example, clicking on a {{htmlelement("label")}} associated with an `appearance: none` checkbox will toggle the checkbox's checked state.

> Because `none` can cause a widget to be hidden, the `base` value is being added to provide widgets with a base appearance, wherein the widget maintains it's native appearance but CSS can be used to change the styles that are not changeable by default. Unlike `none`, in this case the element retains a primitive appearance with default styles that are usable and interoperable.

Reviewer: ```suggestion
The `base` value is being added as an alternative to the `none` value, to guarantee widgets some usable, interoperable native styles as well as a good degree of customization via CSS.
```

Maybe? I found the original really hard to parse.

Author: added your suggestion, but within a full rewrite of the paragraph

> This `base` value is not yet supported, but the `<compat-auto>` values provide similar functionality but are not global in nature. When supported, `base` will reset widgets to their default appearance.

Reviewer: What do you mean by "global" in this paragraph?

> The `base-select` value, which is relevant only for the {{htmlelement("select")}} element and {{cssxref("::picker()", "::picker(select)")}} pseudo-element, enables [styling `<select>` element and the select pickers](#Setting_the_appearance_of_a_select), which represents the part of the form control that pops out of the page, on which the value is applied. The picker pane of {{htmlelements("option")}} elements are rendered in the top-layer, similar to a popover. When `base-select` is set, the picker can be positioned relative to the select (or other element) using CSS anchoring features. When `base-select` is set, the `<select>` does not render outside the browser pane, does not trigger built-in mobile operating system components, and is no longer sized based on the width of the widest `<option>`.

Reviewer: ```suggestion
The `base-select` value, which is relevant only for the {{htmlelement("select")}} element and {{cssxref("::picker()", "::picker(select)")}} pseudo-element, enables [styling `<select>` elements and the select picker](#Setting_the_appearance_of_a_select) (which contains the `<option>` elements). The picker is rendered in the top layer, similar to a popover. When `base-select` is set, the picker can be positioned relative to the select (or other elements) using CSS anchoring features. In addition, the `base-select` value causes the `<select>` to not render outside the browser pane and not trigger built-in mobile operating system components. It is also no longer sized based on the width of the widest `<option>`.
```

### Prefixed non-standard values

> Before standardization, the prefixed **`-moz-appearance`** and **`-webkit-appearance`** allowed elements to be shown as widgets, such as buttons or check boxes. The following non-standard values may be encountered in legacy style sheets, most commonly as values of shadow DOM component [prefixed pseudo-elements](/en-US/docs/Web/CSS/WebKit_Extensions#pseudo-elements):

Reviewer: ```suggestion
Before standardization, the prefixed **`-moz-appearance`** and **`-webkit-appearance`** properties allowed elements to be shown as widgets such as buttons or checkboxes. The following non-standard values may be encountered in legacy style sheets, most commonly as values of shadow DOM component [prefixed pseudo-elements](/en-US/docs/Web/CSS/WebKit_Extensions#pseudo-elements):
```

## Examples

### Basic example

> This example demonstrates altering the appearance of a {{htmlelement("input")}} element using the `appearance` property.

Reviewer: ```suggestion
This example demonstrates altering the appearance of an {{htmlelement("input")}} element using the `appearance` property.
```

#### HTML

We include two `number` form controls along with their labels.

```html
<p>
  <label>Enter a number: <input type="number" min="0" max="10" /></label>
</p>
<p>
  <label
    >Enter a number: <input type="number" min="0" max="10" class="text"
  /></label>
</p>
```

#### CSS

We set the element with the class of `text` to look like a text field.

```css
.text {
  appearance: textfield;
}
```

#### Results

{{EmbedLiveSample("Basic example", 600, 100)}}

> Notice how the spinner is visually removed when the control is set to look like a text field. The `appearance` property has no effect on the functionality: for example, while there is no longer a spinner to click on, the up and down arrows still work.

Reviewer: ```suggestion
Notice how the spinner is visually removed when the control is set to look like a text field. The `appearance` property does not affect the functionality: for example, while there is no longer a spinner to click on, the up and down cursor keys will still increment and decrement the value.
```

Reviewer: Also, in Chrome, the spinner is not gotten rid of using `appearance: textfield`, as far as I can tell. Both look identical?

### Appearance set to `none`

The following example shows how to remove the default styling from a checkbox and {{htmlelement("select")}} element, and apply custom styling.

> The appearance of the checkbox is changed to a circle, and the `<select>` element shows how to remove the arrow indicating that the list can be expanded.

Reviewer: This needs to be rewritten a bit — both the description (the radio buttons are not mentioned, the checkboxes are not changed to look like circles, and the `<select>` appearance is changed a bit more significantly than just removing the expansion arrow) and the code (maybe style the `<select>` a bit differently? Also, the checkbox styling needs a bit more work - when you check a checkbox, the whole layout shifts).

#### HTML

```html
<label><input type="checkbox" /> Default unchecked </label>
<label><input type="checkbox" checked /> Default checked </label>

<hr />

<label><input type="radio" name="radio" /> Default unchecked </label>
<label><input type="radio" name="radio" checked /> Default checked </label>

<hr />

<label
  >Unstyled select
  <select>
    <option>Option 1</option>
    <option>Option 2</option>
  </select>
</label>

<label
  >Styled select
  <select class="none">
    <option>Option 1</option>
    <option>Option 2</option>
  </select>
</label>
```

#### CSS

```css hidden
label {
  display: block;
  margin: 0.5em 0;
}
```

We include styles to apply to both {{htmlelement("input")}} elements of type `checkbox`. We set `appearance: none` to the {{cssxref(":checked")}} UI state for all inputs (`checkbox` and `radio`), as well as to elements with the `.none` class.

```css
[type="checkbox"] {
  width: 1em;
  height: 1em;
  display: inline-block;
  background: red;
}
input:checked,
.none {
  appearance: none;
}
```

#### Result

> {{EmbedLiveSample("Appearance set to none", 600, 200)}}

Reviewer: Example needs to be a bit taller, to get rid of the vertical scrollbar (tested in Firefox)

    <option>Dulce de leche</option>
    <option>Pistachio, rum raisin, and coffee</option>
  </select>
</label>
<label for="ice-cream3"
  >Menulist button flavor:
  <select id="ice-cream3" class="menulistButton">
    <option>Asparagus</option>
    <option>Dulce de leche</option>
    <option>Pistachio, rum raisin, and coffee</option>
  </select>
</label>
```

#### CSS

We select the pickers of all the `<select>` elements using the {{cssxref("::picker()")}} pseudo-element with the `select` parameter. We set the `appearance` value of all the pickers and one `<select>` element to `base-select`. We set the last `<select>` to `menulist-button`. The first `<select>` defaults to the `auto` state:

```css
.baseSelect,
::picker(select) {
  appearance: base-select;
}
.menulistButton {
  appearance: menulist-button;
}
```

```css
label {
  display: block;
}
```

We set values for the `<select>` and pickers' {{cssxref("background-color")}} and {{cssxref("border")}} properties to demonstrate the effects of the `appearance` values:

```css
select {
  border: 1px solid red;
  background-color: orange;
}

::picker(select) {
  background-color: yellow;
  border: none;
}
```

#### Results

> {{EmbedLiveSample("Setting the appearance of a select", 1050, 80)}}

Reviewer: I'd set this to be a bit taller, so that when you open the customizable select, you can see all of the picker.

Author: it's intentional. the picker that is styled is hidden whereas the unstyled ones are visible even though it's outside the iframe.
